chains/substrate/listener/listener.go

```diff
@@ -14,62 +15,67 @@ import (
 )
 
 type EventHandler interface {
-	HandleEvents(evt *events.Events, msgChan chan []*message.Message) error
+	HandleEvents(evts []*events.Events, msgChan chan []*message.Message) error
 }
 type ChainConnection interface {
+	UpdateMetatdata() error
 	GetHeaderLatest() (*types.Header, error)
 	GetBlockHash(blockNumber uint64) (types.Hash, error)
-	GetBlockEvents(hash types.Hash, target interface{}) error
+	GetBlockEvents(hash types.Hash) (*events.Events, error)
+	GetBlockLatest() (*types.SignedBlock, error)
 }
 
 func NewSubstrateListener(connection ChainConnection, eventHandlers []EventHandler, config *substrate.SubstrateConfig) *SubstrateListener {
 	return &SubstrateListener{
 		conn:               connection,
 		eventHandlers:      eventHandlers,
 		blockRetryInterval: config.BlockRetryInterval,
+		blockInterval:      config.BlockInterval,
+		blockConfirmations: config.BlockConfirmations,
 	}
 }
 
 type SubstrateListener struct {
 	conn               ChainConnection
 	eventHandlers      []EventHandler
 	blockRetryInterval time.Duration
+	blockInterval      *big.Int
+	blockConfirmations *big.Int
 }
 
-func (l *SubstrateListener) ListenToEvents(startBlock *big.Int, domainID uint8, blockstore store.BlockStore, stopChn <-chan struct{}, msgChan chan []*message.Message) {
+func (l *SubstrateListener) ListenToEvents(ctx context.Context, startBlock *big.Int, domainID uint8, blockstore store.BlockStore, msgChan chan []*message.Message) {
```

Comment:
This function ListenToEvents is == to EVMListener.ListenToEvents. Although I am not requiring to change something now, it would be good to remind this for future architecture decisions and refactoring.